src/gmtcount.c

```diff
@@ -0,0 +1,935 @@
+	switch (mode) {
+		case GMTCOUNT_MEAN:
+			G->data[ij] += in[GMT_Z];
+			w[kk] += 1.0;
+			break;
+		case GMTCOUNT_MEANW:
+			G->data[ij] += in[GMT_Z] * weight;
+			w[kk] += weight;
+			break;
+		case GMTCOUNT_STD:
+			G->data[ij] += in[GMT_Z];
+			s[kk] += in[GMT_Z] * in[GMT_Z];	/* Sum of z squared */
+			np[kk]++;
+			break;
+		case GMTCOUNT_STDW:
+			G->data[ij] += in[GMT_Z] * weight;	/* Sum of weighted z */
+			w[kk] += weight;	/* Sum of weights */
+			s[kk] += in[GMT_Z] * in[GMT_Z] * weight;	/* Sum of weighted z^2 */
+			np[kk]++;
+			break;
+		case GMTCOUNT_RMS:
+			G->data[ij] += in[GMT_Z] * in[GMT_Z];	/* Sum of z squared */
+			np[kk]++;
+			break;
+		case GMTCOUNT_RMSW:
+			G->data[ij] += in[GMT_Z] * in[GMT_Z] * weight;	/* Sum of weighted z^2 */
+			w[kk] += weight;	/* Sum of weights */
+			break;
+		case GMTCOUNT_MEDIAN: case GMTCOUNT_MODE: case GMTCOUNT_QUANT: case GMTCOUNT_MAD: case GMTCOUNT_LMSSCL: case GMTCOUNT_IRANGE:
+			if (n_alloc[kk] >= np[kk]) {
+				n_alloc[kk] += GMT_SMALL_CHUNK;
+				zp[kk] = gmt_M_memory (GMT, zp[kk], n_alloc[kk], gmt_grdfloat);
+			}
+			zp[kk][np[kk]++] = in[GMT_Z];
+			break;
+		case GMTCOUNT_MEDIANW: case GMTCOUNT_MODEW: case GMTCOUNT_QUANTW: case GMTCOUNT_MADW: case GMTCOUNT_LMSSCLW: case GMTCOUNT_IRANGEW:
+			if (n_alloc[kk] >= np[kk]) {
+				n_alloc[kk] += GMT_SMALL_CHUNK;
+				p[kk] = gmt_M_memory (GMT, p[kk], n_alloc[kk], struct GMT_OBSERVATION);
+			}
+			p[kk][np[kk]].value    = in[GMT_Z];
+			p[kk][np[kk]++].weight = weight;
+			break;
+		case GMTCOUNT_LOWERP:
+			if (in[GMT_Z] <= 0.0) return;	/* Only consider positive values */
+			/* Fall through on purpose */
+		case GMTCOUNT_LOWER:
+			if (visited[kk]) {	/* Been here before, so can make a comparison */
+				if (in[GMT_Z] < G->data[ij]) G->data[ij] = in[GMT_Z];
+			}
+			else {	/* First time, mark it */
+				G->data[ij] = in[GMT_Z];
+				visited[kk] = 1;
+			}
+			break;
+		case GMTCOUNT_UPPERP:
+			if (in[GMT_Z] >= 0.0) return;	/* Only consider negative values */
+			/* Fall through on purpose */
+		case GMTCOUNT_UPPER:
+			if (visited[kk]) {	/* Been here before, so can make a comparison */
+				if (in[GMT_Z] > G->data[ij]) G->data[ij] = in[GMT_Z];
+			}
+			else {	/* First time, mark it */
+				G->data[ij] = in[GMT_Z];
+				visited[kk] = 1;
+			}
+			break;
+		case GMTCOUNT_ZRANGE:
+			if (visited[kk]) {	/* Been here before, so can make a comparison */
+				if (in[GMT_Z] > G->data[ij]) G->data[ij] = in[GMT_Z];
+				if (in[GMT_Z] < s[kk]) s[kk] = in[GMT_Z];
+			}
+			else {	/* First time, mark it */
+				G->data[ij] = s[kk] = in[GMT_Z];
+				visited[kk] = 1;
+			}
+			break;
+		case GMTCOUNT_SUM: case GMTCOUNT_COUNTW:
+			G->data[ij] += in[GMT_Z];
+			break;
+		default:	/* count */
+			G->data[ij] += 1.0;
+			break;
+	}
+}
+
+bool outside_hexagon (struct GMT_CTRL *GMT, double x0, double y0, double x, double y, double distance, double radius_inner) {
```

Comment:
This is a bool function returning an int. I had recently hard to find bug in a similar situation and VS2019.

The binning could perhaps be speed up if points are first checked to fall inside the circumscribed rectangle. Also remember that each **GMT_Report()** is an *if* branch, and there seems to be one for each input data point.

Reply:
OK, fixing the bool.  The GMT_Report was mostly there for me during debugging.  I can place them under #ifdef DEBUG since just there for keeping.
I will look at **gmtselect** for the pre-circle rectangular skip.  But implementing -Tr for rectangles anyway.  For geographic we can do a check on latitude, but longitude by itself cannot work with -S.

Reply:
latitude checks for rectangle is also tricky across poles.